Source/WebKit/Platform/IPC/Connection.cpp

```diff
@@ -1005,7 +1034,7 @@ void Connection::dispatchSyncMessage(Decoder& decoder)
         RELEASE_ASSERT(unwrappedDecoder);
         processIncomingMessage(WTFMove(unwrappedDecoder));
 
-        SyncMessageState::singleton().dispatchMessages();
+        syncState().dispatchMessages();
     } else {
         // Hand off both the decoder and encoder to the client.
```

Comment:
in Connection::invalidate() you need to add a note 
// Note: currently we bind the sync state to main runloop, which will always exist. As such, we do not yet disconnect the connection from it, even though in reality we should not dispatch through the runloop after invalidate has happened.

Reply:
But then you might note that you need the explicit release operation.
Then you have the option where should that happen: in the wrong place (~Connection) or in the right place (Connection::invalidate). If you then choose the right place, then you need to prove that the ptr is not accessed in  a wrong way from receive thread nor from the submitted closures. Which then leads to some other changes..